config/m4/compiler.m4

```diff
@@ -17,8 +17,26 @@ AC_ARG_ENABLE(debug,
         [],
         [enable_debug=no])
 AS_IF([test "x$enable_debug" == xyes],
-        [BASE_CFLAGS="-O0 -D_DEBUG $BASE_CFLAGS"],
-        [BASE_CFLAGS="-O3 $BASE_CFLAGS"])
+        [BASE_CFLAGS="-D_DEBUG $BASE_CFLAGS"],
+        [])
+AS_IF([test "x$enable_debug" == xyes],
+        [AS_IF([test "$enable_compiler_opt" == none], [enable_compiler_opt="0"], [])],
+        [])
+
+#
+# Optimization level
+#
+AC_ARG_ENABLE(compiler-opt,
+        AC_HELP_STRING([--enable-compiler-opt], [Set optimization level [0-3]]),
+        [],
+        [enable_compiler_opt="none"])
+AS_IF([test "$enable_compiler_opt" == "yes"], [BASE_CFLAGS="-O3 $BASE_CFLAGS"],
+      [test "$enable_compiler_opt" == "none"],
```

Comment:
why do we need special case for 'none'? seems like we can say that the default value of "enable_compiler_opt" can be "yes" rather than none. Since in both cases, "enable_debug=yes" would override "enable_compiler_opt" to "-O0"

Reply:
'none' is special case when user didn't set any value, in this case it sets optimization level based on debug configuration. in case if user set level (```--enable-compiler-opt``` or ```--enable-compiler-opt=value```) system forces specified level. The "feature" here is set O3 for release and O0 for debug by default, but force custom level in case if user wants it

Reply:
so why is the 'if' in line 22 needed?

Reply:
artefact (to prevent out-of-ordering variable settings)
removed